core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlPrepareAsyncProcessor.java

```diff
@@ -21,36 +21,44 @@
 import com.datastax.oss.driver.api.core.type.reflect.GenericType;
 import com.datastax.oss.driver.internal.core.context.InternalDriverContext;
 import com.datastax.oss.driver.internal.core.session.DefaultSession;
-import com.datastax.oss.driver.internal.core.session.RequestHandler;
 import com.datastax.oss.driver.internal.core.session.RequestProcessor;
-import java.nio.ByteBuffer;
+import com.datastax.oss.driver.internal.core.util.concurrent.CompletableFutures;
+import com.datastax.oss.driver.shaded.guava.common.cache.Cache;
+import com.datastax.oss.driver.shaded.guava.common.cache.CacheBuilder;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
-import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
 import net.jcip.annotations.ThreadSafe;
 
 @ThreadSafe
 public class CqlPrepareAsyncProcessor
     implements RequestProcessor<PrepareRequest, CompletionStage<PreparedStatement>> {
 
-  private final ConcurrentMap<ByteBuffer, DefaultPreparedStatement> preparedStatementsCache;
-
-  public CqlPrepareAsyncProcessor(
-      ConcurrentMap<ByteBuffer, DefaultPreparedStatement> preparedStatementsCache) {
-    this.preparedStatementsCache = preparedStatementsCache;
-  }
+  private final Cache<PrepareRequest, CompletableFuture<PreparedStatement>> cache =
+      CacheBuilder.newBuilder().weakValues().build();
```

Comment:
What if the user wants to customize the cache? It doesn't seem easy to pass a `DriverContext` when instantiating processors.

Reply:
I don't think customizing the cache will be a very common need, its size/eviction strategy/etc is mainly driven by how your application uses prepared statements.

On the other hand it doesn't hurt to expose a protected constructor that takes the cache as a parameter, so I'll do that.